model: Add eager-embed embedding model

jpbalarini opened the pull request:
Add inference code for eager-embed embedding model.
[eager-embed-v1](https://huggingface.co/eagerworks/eager-embed-v1) is a multimodal dense embedding model with a 2560 embed dimension based on Qwen3-VL and finetuned on multiple public datasets.

More info here:
https://huggingface.co/eagerworks/eager-embed-v1
https://github.com/eagerworks/eager-embed

**Checklist**:
- [x] I have filled out the ModelMeta object to the extent possible
- [x] I have ensured that my model can be loaded using
  - [x] `mteb.get_model(model_name, revision)` and
  - [x] `mteb.get_model_meta(model_name, revision)`
- [x] I have tested the implementation works on a representative set of tasks.
- [x] The model is public, i.e. is available either as an API or the weight are publicly available to download

KennethEnevoldsen commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+        )
+        requires_package(
+            self, "qwen_vl_utils", model_name, "pip install mteb[qwen_vl_utils]"
+        )
```
I would convert the optional dependencies to be model specific (so one group installs everything needed)

KennethEnevoldsen commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+            if len(text_embeddings) != len(image_embeddings):
+                raise ValueError(
+                    "The number of texts and images must have the same length"
+                )
```
this is not a requirement on our end (one text can contain multiple images) - if you want to keep this then I would change the message to specifically refer to the (this model does not support ... or this implementation of the model does not suport) to avoid confusion.

jpbalarini replied:
I took inspiration from here https://github.com/embeddings-benchmark/mteb/blob/main/mteb/models/model_implementations/jina_models.py#L329
but changed the message to what you suggested

KennethEnevoldsen reviewed (commented):
A few minor comments - otherwise the submission looks good

Samoed commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+            text_embeddings = self.get_text_embeddings(inputs, **kwargs)
+
+        if "image" in inputs.dataset.features:
+            image_embeddings = self.get_image_embeddings(inputs, **kwargs)
```
Probably your model can encode them both

Samoed commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+                query_texts = [
+                    self.processor.apply_chat_template(
+                        msg, tokenize=False, add_generation_prompt=False
+                    ) + "<|endoftext|>"
```
Not sure about adding `"<|endoftext|>"`

jpbalarini commented on pyproject.toml:
```diff
@@ -93,6 +93,7 @@ xet = ["huggingface_hub>=0.32.0"]
 youtu = ["tencentcloud-sdk-python-common>=3.0.1454", "tencentcloud-sdk-python-lkeap>=3.0.1451"]
 llama-embed-nemotron = ["transformers==4.51.0"]
 faiss-cpu = ["faiss-cpu>=1.12.0"]
+qwen_vl_utils = ["qwen_vl_utils>=0.0.14"]
```
done!

jpbalarini commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+            kwargs['dtype'] = kwargs.pop('torch_dtype')
+
+        # Load model
+        if self.use_peft:
```
I was using it while finetuning the model, but I removed it now.

jpbalarini commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+            batch_size=batch_size,
+            collate_fn=text_collate_fn,
+            shuffle=False,
+        )
```
agreed, removed it

jpbalarini commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+            batch_size=batch_size,
+            collate_fn=image_collate_fn,
+            shuffle=False,
+        )
```
removed it

jpbalarini commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+                        {
+                            'role': 'user',
+                            'content': [
+                                {'type': 'text', 'text': f'Query: {query}'},
```
added

jpbalarini commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,334 @@
+                        {
+                            'role': 'user',
+                            'content': [
+                                {'type': 'text', 'text': f'Query: {query}'},
```
added

Samoed commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,286 @@
+         **kwargs,
+    ):
+        """Encode texts (queries) into embeddings."""
+        from qwen_vl_utils import process_vision_info
```
I think `process_vision_info` wouldn't do anything in this code, and you can remove `qwen_vl_utils` dependency https://github.com/QwenLM/Qwen3-VL/blob/b6b80717cb28887eb80e82fc96239c1c53a84cc6/qwen-vl-utils/src/qwen_vl_utils/vision_process.py#L501

jpbalarini commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,286 @@
+
+        # Load processor
+        self.processor = AutoProcessor.from_pretrained(model_name)
+        self.processor.tokenizer.padding_side = "left"
```
good point

Samoed commented:
@jpbalarini Did you try to encode images and texts together without separation on image/text modalities

Samoed commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,283 @@
+                    for msg in doc_messages
+                ]
+
+                doc_image_inputs, doc_video_inputs = process_vision_info(doc_messages)
```
I still think that you don't need `process_vision_info`, because `image_processor` will resize images too https://github.com/huggingface/transformers/blob/672dc07527e10ce8197b0756527d99ee622ff24d/src/transformers/models/qwen2_vl/image_processing_qwen2_vl.py#L271-L280

Samoed commented:
@jpbalarini I've added a fix https://github.com/embeddings-benchmark/mteb/pull/3618. Thank you for reporting!

jpbalarini commented:
> @jpbalarini I've added a fix #3618. Thank you for reporting!

You're welcome! Let me add the latest changes and rerun the benchmark to see that everything works as expected

Samoed commented on mteb/tasks/retrieval/multilingual/vidore2_bench_retrieval.py:
```diff
@@ -41,7 +41,11 @@ def _load_data(
             },
             remove_columns=["query-id", "query"],
         )
-        query_ds = query_ds.select_columns(["id", "text"])
```
Yes, my bad. You can just remove this line then

Samoed commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,196 @@
+        """Calculate cosine similarity between embeddings."""
+        return torch.nn.functional.cosine_similarity(
+            a.unsqueeze(1), b.unsqueeze(0), dim=-1
+        )
```
`calculate_probs` won't be used and `similarity` will be used automatically
```suggestion
```

Samoed commented on mteb/tasks/retrieval/multilingual/vidore2_bench_retrieval.py:
```diff
@@ -41,7 +41,11 @@ def _load_data(
+        if langs is None:
+            query_ds = query_ds.select_columns(["id", "text"])
+        else:
+            query_ds = query_ds.select_columns(["id", "text", "language"])
```
You can just change later in code
```diff
if langs is None:
-    queries[split] = query_ds
+    queries[split] = query_ds.select_columns(["id", "text"])
    corpus[split] = corpus_ds
```

Samoed commented on mteb/models/model_implementations/eagerworks_models.py:
```diff
@@ -0,0 +1,168 @@
+Eager_Embed_V1 = ModelMeta(
+    loader=EagerEmbedV1Wrapper,
+    loader_kwargs=dict(
+        torch_dtype=torch.float16,
```
```suggestion
        dtype=torch.float16,
```

jpbalarini replied:
good catch

jpbalarini commented:
> @jpbalarini Is this the final version? Have you submitted all the results with the processing of text and images combined?

Yes @Samoed I updated the new results [here](https://github.com/embeddings-benchmark/results/pull/328) (and I added vidore v3 too).
It should be ready for merging! Thanks for all the comments
